conda-store-server/conda_store_server/action/download_packages.py

```diff
@@ -28,6 +42,29 @@ def action_fetch_and_extract_conda_packages(
             file_path = pkgs_dir / filename
             count_message = f"{packages_searched} of {total_packages}"
             with filelock.FileLock(str(lock_filename)):
+                # This magic file, which is currently set to "urls.txt", is used
+                # to check cache permissions in conda, see _check_writable in
+                # PackageCacheData.
+                #
+                # Sometimes this file is not yet created while this action is
+                # running. Without this magic file, PackageCacheData cache
+                # query functions, like query_all, will return nothing.
+                #
+                # If the magic file is not present, this error might be thrown
+                # during the lockfile install action:
+                #
+                # File "/opt/conda/lib/python3.10/site-packages/conda/misc.py", line 110, in explicit
+                #   raise AssertionError("No package cache records found")
+                #
+                # The code below is from create_package_cache_directory in
+                # conda, which creates the package cache, but we only need the
+                # magic file part here:
+                cache_magic_file = pkgs_dir / PACKAGE_CACHE_MAGIC_FILE
+                if not cache_magic_file.exists():
```

Comment:
This might still allow for a race condition, doesn't it? `try-except` constructions work better here. I wonder if we should fix this upstream too.

Reply:
`if exists` might be prone to race conditions in some cases (but not always, see below). It's a classic [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) issue. `try/except` might be racy as well. It depends on where and how that's used. The general idea: you try using the shared resource right away without doing any checks and handle any failures if any later.

That said, I think it's actually fine to keep `if exists` here because I'm _assuming_ this file is never deleted once it's created. So this check is just to:
- avoid modifying file metadata such as creation/modification time for no reason (this might be useful later to understand when the cache was created or modified)
- avoiding unnecessary work and making it faster (I'm assuming the syscall that checks whether the file exists returns faster compared to the one that needs to populate the metadata).

No fix is needed upstream either because `create_package_cache_directory` has `try/except`. So it always calls `touch` on the file first before potentially throwing an error. Note, however, that the usecase is different here. I'm assuming `create_package_cache_directory` is called only once when initializing the cache while the code above is called for every build, so I want to avoid unnecessary work.